torchaudio/functional.py

```diff
@@ -844,6 +844,163 @@ def compute_deltas(specgram, win_length=5, mode="replicate"):
+           bound (Default: `1.0`)
+           Example: interval=1.0 -> [-1.0, 1.0]
+    Returns:
+       torch.Tensor: waveform scaled to interval
+    """
+    abs_max = torch.max(torch.abs(waveform))
+    ratio = abs_max / interval
+    waveform /= ratio
+
+    return waveform
+
+
+def _add_noise_shaping(dithered_waveform, waveform):
+    r"""Noise shaping is calculated by error:
+    error[n] = dithered[n] - original[n]
+    noise_shaped_waveform[n] = dithered[n] + error[n-1]
+    """
+    error = dithered_waveform - waveform
+
+    # add error[n-1] to dithered_waveform[n], so offset the error by 1 index
+    for index in range(error.size()[0]):
+        err = error[index]
+        error_offset = torch.cat((torch.zeros(1), err))
+        error[index] = error_offset[:waveform.size()[1]]
+
+    noise_shaped = dithered_waveform + error
+    return noise_shaped
+
+
+@torch.jit.script
+def probability_distribution(waveform, density_function="TPDF"):
+    # type: (Tensor, str) -> Tensor
+    r"""Perform a probability distribution function on a waveform.
+
+    Triangular probability density function (TPDF) dither noise has a 
+    triangular distribution; values in the center of the range have a higher 
+    probability of occurring. 
+
+    Rectangular probability density function (RPDF) dither noise has a 
+    uniform distribution; any value in the specified range has the same
+    probability of occurring.
+
+    Gaussian probability density function (GPDF) has a normal distribution.
+    The relationship of probabilities of results follows a bell-shaped,
+    or Gaussian curve, typical of dither generated by analog sources.
+    Args:
+        waveform (torch.Tensor): Tensor of audio of dimension (channel, time)
+        probability_density_function (string): The density function of a
+           continuous random variable (Default: `TPDF`)
+           Options: Triangular Probability Density Function - `TPDF`
+                    Rectangular Probability Density Function - `RPDF`
+                    Gaussian Probability Density Function - `GPDF`
+    Returns:
+        torch.Tensor: waveform dithered with TPDF
+    """
+    wave_size = waveform.size()[1] - 1
+    number_of_bits = 16
+
+    up_scaling = 2 ** (number_of_bits - 1) - 2
+    signal_scaled = waveform * up_scaling
+    down_scaling = 2 ** (number_of_bits - 1)
+
+    distributed = waveform
+    if (density_function == "RPDF"):
+        random_tensor = int(torch.randint(wave_size, [1, ]).item())
+        RPDF = waveform[0][random_tensor] - 0.5
+
+        signal_scaled_RPDF = signal_scaled + RPDF
+        quantised_signal_scaled_RPDF = torch.round(signal_scaled_RPDF)
+        quantised_signal_RPDF = quantised_signal_scaled_RPDF / down_scaling
+
+        distributed = quantised_signal_RPDF
+    elif (density_function == "GPDF"):
+        gaussian = torch.distributions.normal.Normal(torch.mean(waveform), 1).sample()
```

Comment:
This is the `torch.jit.frontend.NotSupportedError: comprehension ifs not supported yet:` error. Should I wait on that to be fixed, or remove the JIT tags for these functions? (see [issue #29843](https://github.com/pytorch/pytorch/issues/29843))

Reply:
Oh right. While waiting for this to be implemented, we discussed two options, no?

1. Comment out the jit decoration 
```
# TODO Add once https://github.com/pytorch/pytorch/issues/29843 is resolved
# @torch.jit.script
```

2. Use the alternative that didn't rely on `torch.distributions`.
```
# TODO Replace by distribution code once https://github.com/pytorch/pytorch/issues/29843 is resolved
# gaussian = torch.distributions.normal.Normal(torch.mean(waveform), 1).sample()
```

Option 1 is acceptable for this PR since it's a new feature and we can add jit support in stage. Option 2 is preferrable, since it doesn't sacrifice functionality, but takes a little more effort now and in maintenance. Thoughts?

Reply:
Implementing option 2!